### ISSUE TITLE
`invalid_reference_casting` lint catches some patterns that aren't UB

### ISSUE THREAD

**Issue by LegionMammal978** [rust-lang/rust#116410]
When reading the 1.73.0 release notes, I noticed that the `invalid_reference_casting` lint was made deny-by-default in #112431, under the premise that casting `&T` to `&mut T` via raw pointers is always UB under Stacked Borrows. However, the lint appears a bit overly strict, since there are two cases where such a cast does not cause UB:
```rust
#![allow(invalid_reference_casting)]
use std::cell::UnsafeCell;

fn main() {
    // 1. Every byte in `T` is inside an `UnsafeCell`:
    {
        let r1: &UnsafeCell<i32> = &UnsafeCell::new(0);
        let r2: &mut i32 = unsafe { &mut *(r1 as *const _ as *mut _) };
        *r2 = 42;
    }
    // 2. `T` is a zero-sized type:
    {
        let r1: &() = &();
        let r2: &mut () = unsafe { &mut *(r1 as *const _ as *mut _) };
        *r2 = ();
    }
}
```

Without `#![allow(invalid_reference_casting)]`, both of these casts trigger the lint in 1.75.0-nightly (187b8131d 2023-10-03):
```rust
error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
 --> src/main.rs:8:37
  |
8 |         let r2: &mut i32 = unsafe { &mut *(r1 as *const _ as *mut _) };
  |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
  = note: `#[deny(invalid_reference_casting)]` on by default

error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
  --> src/main.rs:14:36
   |
14 |         let r2: &mut () = unsafe { &mut *(r1 as *const _ as *mut _) };
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
```

However, when the lint is disabled, Miri runs the program successfully, under both Stacked Borrows and Tree Borrows. This makes sense looking at the rules. In the first case, the tag of `r1` (as well as `r1 as *const _`) has `SharedReadWrite` permission to all bytes of the `i32`, since the memory is inside an `UnsafeCell`. Therefore, it is able to perform a write access that grants `Unique` permission to the new `&mut i32`. In the second case, the reference doesn't point to any bytes, so it doesn't need to perform any access or be granted any permissions.

(In the `UnsafeCell` case under Tree Borrows, `r1` is `Reserved` when it is created, to match the permissions of the temporary it comes from. `r2` is also `Reserved` when it is created; when it is assigned to, `r2` becomes `Active`, and `r1` also becomes `Active` due to the child write. In the ZST case, the zero-size accesses when creating and assigning to `r2` still do not need any permissions.)

Both of these patterns look a bit dubious when expressed as a reference cast, but neither of them are UB. (At worst, the `UnsafeCell` case is somewhat dubious for accessing an `UnsafeCell` mutably without using `UnsafeCell::get()`.) Therefore, I would suggest downgrading or turning off the lint when all bytes of the destination type are inside an `UnsafeCell` in the source type. If this criterion is impractical to implement, it could be reduced to either the source type being `!Freeze` or the destination type being a ZST, at the cost of introducing a few false negatives.

**Comment by Urgau** [comment]
There is a difference between library UB and language UB and this is an example of one.

The documentation of `UnsafeCell` [specifically says](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html#memory-layout):
> Note that the **only valid way** to obtain a `*mut T` pointer to the contents of a shared UnsafeCell<T> is through **[.get()](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html#method.get)** or **[.raw_get()](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html#method.raw_get)**.

And even calls out against what your describing:
> Even though `T` and `UnsafeCell<T>` have the same memory layout, the following is **not allowed** and **undefined behavior**:
>
> ```rust
> unsafe fn not_allowed<T>(ptr: &UnsafeCell<T>) -> &mut T {
>    let t = ptr as *const UnsafeCell<T> as *mut T;
>    // This is undefined behavior, because the `*mut T` pointer
>    // was not obtained through `.get()` nor `.raw_get()`:
>    unsafe { &mut *t }
>}
> ```


@RalfJung said [here](https://github.com/rust-lang/rust/pull/112431#discussion_r1255803991) that he felt that linting here was justified. That still the case?

@rustbot label -needs-triage +A-lint

**Comment by RalfJung** [comment]
Yes, that was deliberate.  But of course if people find it confusing we can always reconsider.

**Comment by LegionMammal978** [comment]
> There is a difference between library UB and language UB and this is an example of one.
> 
> The documentation of `UnsafeCell` [specifically says](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html#memory-layout):
> 
> > Note that the **only valid way** to obtain a `*mut T` pointer to the contents of a shared UnsafeCell is through **[.get()](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html#method.get)** or **[.raw_get()](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html#method.raw_get)**.
> 
> And even calls out against what your describing:
> 
> > Even though `T` and `UnsafeCell<T>` have the same memory layout, the following is **not allowed** and **undefined behavior**:
> > ```rust
> > unsafe fn not_allowed<T>(ptr: &UnsafeCell<T>) -> &mut T {
> >    let t = ptr as *const UnsafeCell<T> as *mut T;
> >    // This is undefined behavior, because the `*mut T` pointer
> >    // was not obtained through `.get()` nor `.raw_get()`:
> >    unsafe { &mut *t }
> > }
> > ```
> 
> To me the lint is working as intended, this isn't a bug.

I see how the first cast causes library UB according to the `UnsafeCell` documentation. However, a modified version of the first cast is still forbidden by the lint, even though it correctly uses `.get()` to "obtain a `*mut T` pointer to the contents of a shared `UnsafeCell<T>`":
```rust
let r1: &UnsafeCell<i32> = &UnsafeCell::new(0);
let r2: &mut UnsafeCell<i32> = unsafe { &mut *(r1 as *const _ as *mut _) };
unsafe { *r2.get() = 42 };
```

Also, even without `UnsafeCell`, there's still the case where the destination type being a ZST and requires no permission, but the lint still catches it. It's permitted to create a `&mut ZST` reference from any (non-null aligned) address at all, except for those of dangling allocations and locals, and creating it from an immutable reference seems like just another case of that. Unless we mean to have some secret unwritten language UB that prevents `&ZST`-to-`&mut ZST` casts?

**Comment by Urgau** [comment]
> ```rust
> let r1: &UnsafeCell<i32> = &UnsafeCell::new(0);
> let r2: &mut UnsafeCell<i32> = unsafe { &mut *(r1 as *const _ as *mut _) };
> unsafe { *r2.get() = 42 };
> ```

`r2` is doing exactly what the documentation is saying to be undefined behaviour. You want the `.get()` to be done on the creation of `r2`, like this:

```rust
let r1: &UnsafeCell<i32> = &UnsafeCell::new(0);
let r2: &mut UnsafeCell<i32> = unsafe { &mut *(r1.get() as *mut _) };
```

Otherwise it's still undefined behaviour according to the documentation.

As for the ZST concern I will let someone more knowledgeable address it.

**Comment by LegionMammal978** [comment]
> > ```rust
> > let r1: &UnsafeCell<i32> = &UnsafeCell::new(0);
> > let r2: &mut UnsafeCell<i32> = unsafe { &mut *(r1 as *const _ as *mut _) };
> > unsafe { *r2.get() = 42 };
> > ```
> 
> `r2` is doing exactly what the documentation is saying to be undefined behavior. You want the `.get()` to be done on the creation of `r2`, like this:
> 
> ```rust
> let r1: &UnsafeCell<i32> = &UnsafeCell::new(0);
> let r2: &mut UnsafeCell<i32> = unsafe { &mut *(r1.get() as *mut _) };
> ```
> 
> Otherwise it's still undefined behaviour according to the documentation.

Where in the documentation does it talk "exactly" about creating a `&mut UnsafeCell<T>` from an `&UnsafeCell<T>`? My point is that the quoted text is talking only about creating a pointer to the *contents* of the `UnsafeCell<T>`, and says nothing about creating one to the *cell itself*.

---

If we go through the literal text of the [`UnsafeCell`](https://doc.rust-lang.org/1.72.1/std/cell/struct.UnsafeCell.html) documentation regarding `&mut` references, it says:

> If you have a reference `&T`, then normally in Rust the compiler performs optimizations based on the knowledge that `&T` points to immutable data. Mutating that data, for example through an alias or by transmuting an `&T` into an `&mut T`, is considered undefined behavior.

This is specifically referring to the case where `T` does not have interior mutability, and `&T` is truly immutable. (However, it is still missing the caveat of ZST references, unless we stretch "normally" to exclude that.)

> There is no legal way to obtain aliasing `&mut`, not even with `UnsafeCell<T>`.

`r2` is not aliased with any other `&mut` references.

> If you create a safe reference with lifetime `'a` (either a `&T` or `&mut T` reference), then you must not access the data in any way that contradicts that reference for the remainder of `'a`.

A write access doesn't contradict an `&UnsafeCell<T>` reference.

> A `&mut T` reference may be released to safe code provided neither other `&mut T` nor `&T` co-exist with it. A `&mut T` must always be unique.

`r2` remains unique for its lifetime, and indeed it could be safely passed into a callback (as long as that callback can't cause a read access on `r1`).

> Note that whilst mutating the contents of an `&UnsafeCell<T>` (even while other `&UnsafeCell<T>` references alias the cell) is ok (provided you enforce the above invariants some other way), it is still undefined behavior to have multiple `&mut UnsafeCell<T>` aliases. That is, `UnsafeCell` is a wrapper designed to have a special interaction with *shared* accesses (*i.e.*, through an `&UnsafeCell<_>` reference); there is no magic whatsoever when dealing with *exclusive* accesses (*e.g.*, through an `&mut UnsafeCell<_>`): neither the cell nor the wrapped value may be aliased for the duration of that `&mut` borrow.

Once again, `r2` is not aliased with any other `&mut UnsafeCell<i32>` references. Perhaps this wording could be stretched as the cast "losing" the magic and making `r2.get()` no longer work, but one might reasonably think that the implicit `&*r2` reborrow would then "regain" the magic.

> Note that the only valid way to obtain a `*mut T` pointer to the contents of a *shared* `UnsafeCell<T>` is through [`.get()`](https://doc.rust-lang.org/1.72.1/std/cell/struct.UnsafeCell.html#method.get) or [`.raw_get()`](https://doc.rust-lang.org/1.72.1/std/cell/struct.UnsafeCell.html#method.raw_get). A `&mut T` reference can be obtained by either dereferencing this pointer or by calling `.get_mut()` on an *exclusive* `UnsafeCell<T>`.

This is the main relevant passage here. Again, this is very specifically talking about "obtain[ing] a `*mut T` pointer to the contents". When I convert `r1` to `r2`, I am not obtaining a mutable reference to the `i32` contained in the cell, but to the `UnsafeCell<i32>` itself. Then, I convert that to a shared reference and call `.get()` to validly obtain a `*mut i32` pointer to the contents. At no point do I create a mutable pointer to the contents without calling `.get()` or `.raw_get()`.

**Comment by Urgau** [comment]
Sorry, I didn't realise your point was about `&UnsafeCell<T> -> &mut UnsafeCell<T>` regardless of the inner `T`.

That's an interesting point.

The reference seems to agree with your interpretation:
> [Behavior considered undefined](https://doc.rust-lang.org/stable/reference/behavior-considered-undefined.html#behavior-considered-undefined): 
> Mutating immutable data. All data inside a [const](https://doc.rust-lang.org/stable/reference/items/constant-items.html) item is immutable. Moreover, all data reached through a shared reference or data owned by an immutable binding is immutable, **unless that data is contained within an [UnsafeCell<U>](https://doc.rust-lang.org/stable/std/cell/struct.UnsafeCell.html)**.

That would means that:
 - `&UnsafeCell<T> -> &mut UnsafeCell<T>` is fine
 - `&T -> &mut T` is only fine, if `T` has interior mutability
 - `&UnsafeCell<T> -> &mut T` is not fine and should use `UnsafeCell::get` or `UnsafeCell::raw_get`

Does this make sense?
This would means that for `&A -> &mut B` to be considered UB with casts:
  - `A` should not have interior mutability
  - OR if `A` has interior mutability then `B` should not have interior mutability (couldn't call `UnsafeCell::{raw_,}get`)

**Comment by RalfJung** [comment]
> &UnsafeCell<T> -> &mut UnsafeCell<T> is fine

Eh what? `&T -> &mut T` is basically never fine, for any `T`. This has nothing to do with `UnsafeCell`.

**Comment by Urgau** [comment]
> > &UnsafeCell -> &mut UnsafeCell is fine
> 
> Eh what? `&T -> &mut T` is basically never fine, for any `T`. This has nothing to do with `UnsafeCell`.

My bad, I mixed-up some things in my head, opsem is really not my strong suite. Sorry for the noise.

**Comment by LegionMammal978** [comment]
> > &UnsafeCell -> &mut UnsafeCell is fine
> 
> Eh what? `&T -> &mut T` is basically never fine, for any `T`. This has nothing to do with `UnsafeCell`.

The thing is, we *never actually say anywhere* that `&UnsafeCell<T>` → `&mut UnsafeCell<T>` is not fine. The documentation for `UnsafeCell` specifically says that `&T` → `&mut T` is not fine *when `&T` is a reference to immutable data*, and that `&UnsafeCell<T>` is the exception to this rule, since it allows the data to be mutable.

Then, later, it expressly forbids converting `&UnsafeCell<T>` to `&mut T` directly, saying that the only valid way to mutably access the *contents* is through `.get()` or `.raw_get()`. However, the paragraph forbidding that says absolutely nothing about converting `&UnsafeCell<T>` to `&mut UnsafeCell<T>`. And nothing outside of that paragraph forbids that conversion either: the original `&UnsafeCell<T>` is valid for write accesses in general, and there's nothing violating the uniqueness of the `&mut UnsafeCell<T>` derived from it.

**Comment by RalfJung** [comment]
When it comes to library docs, everything that is not allowed is forbidden. You cannot make any assumptions about the internals of a library type except for those that are explicitly documented.

**Comment by RalfJung** [comment]
> the original &UnsafeCell<T> is valid for write accesses in general,

No, it isn't. At least we haven't documented that. The only thing we have documented is that you can call `get` on it *and write to that pointer*. For all you know, `&UnsafeCell` is a read-only pointer that only becomes read-write in `get`.

That's not how Stacked Borrows or Tree Borrows work, but it is possible (not likely, but possible) that we end up with an aliasing model that works like that.

**Comment by LegionMammal978** [comment]
> When it comes to library docs, everything that is not allowed is forbidden. You cannot make any assumptions about the internals of a library type except for those that are explicitly documented.

Isn't there general agreement that certain assumptions *are* allowed, even when the standard library docs don't explicitly permit them? In other words, the set of permitted unsafe operations on arbitrary types is strictly larger than "operations that can be performed with safe code and the standard library". The archetypical example here is `take_mut`, which [you consider to be sound](https://github.com/rust-lang/rfcs/pull/1736#issuecomment-1311571940), and so do I, even though it has observable effects that cannot be generated safely. Until we have a grand unified theory of safety invariants, there is a gray area wherever the docs don't explicitly permit nor forbid a particular operation.

**Comment by RalfJung** [comment]
We're grandfathering in patterns like take_mut, but this approach is in general not tenable -- it is not hard to end up in situations where two such extensions are mutually incompatible.

Also, this issue here is about an opsem questions more than it is about a safety invariant question.

**Comment by LegionMammal978** [comment]
How does #116421 solve the case of `&ZST`-to-`&mut ZST` casts, to which the current aliasing models do not ascribe any language UB? This issue is **not** just about `UnsafeCell`. The restriction on ZSTs, including user-defined ZSTs to which no library UB applies, causes the compiler to deny certain patterns perfectly acceptable in the present "SB Rust" or "TB Rust", and I do not recall a precedent of denying-by-default patterns that may or may not be language UB in the future.

**Comment by RalfJung** [comment]
For gray area we generally do reserve the right to deny, I think, in particular if there are good alternatives.

But yes for ZST the argument could be made that we're unlikely to make this actually UB in the future so maybe we shouldn't lint against this. I don't have a strong opinion either way, just a technical concern -- it can be hard to find out if the type is a ZST. We could possibly silence the lint if we're *sure* it's a ZST but still lint for generic types that *might* be ZST. The only concern with that is that the fact that the lint disappears on ZST could be interpreted as a promise that this is fine...

**Comment by LegionMammal978** [comment]
> But yes for ZST the argument could be made that we're unlikely to make this actually UB in the future so maybe we shouldn't lint against this. I don't have a strong opinion either way, just a technical concern -- it can be hard to find out if the type is a ZST. We could possibly silence the lint if we're sure it's a ZST but still lint for generic types that might be ZST. The only concern with that is that the fact that the lint disappears on ZST could be interpreted as a promise that this is fine...

Sure, I have no qualms with adding extra restrictions on "private ZST" types and whatnot, akin to the restrictions on ZST fields for `repr(transparent)`. My main concern is that deny-by-default lints (except for future-incompat lints) should only be used for patterns that can't and won't ever do what they look like they're doing, and I don't think that performing odd but legal aliasing operations according to the current models meet this bar. (After all, what if we want to write programs specifically to demonstrate the minutiae of the aliasing models?)

If implicit promises are a real concern, then perhaps it might be useful to have something less forceful than a deny-by-default lint for patterns that could be dubious under alternative conceptions of the aliasing rules. But I personally don't see it as that big of an issue, given how compiler UB detection has always explicitly been on a "best-effort" basis rather than a hard guarantee.

Regardless, can this issue be re-opened until the ZST question is resolved?

**Comment by RalfJung** [comment]
> odd but legal aliasing operations according  according to the current models 

We don't have a "current model", we just have a bunch of non-normative experiments.

But anyway, I reopened. Can you update the issue description to reflect that this is only about the ZST case now?

**Comment by LegionMammal978** [comment]
I have updated the issue description.
